From 7e0cc0fd6aa2c035f5c4da71735046d158a2219d Mon Sep 17 00:00:00 2001
From: Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>
Date: Thu, 21 Jul 2022 14:54:07 +0300
Subject: [PATCH 14/21] libxl: Check for d_config to be valid in
 device_model_spawn_outcome()

Check that d_config is not NULL before dereferencing it. For instance,
this is a case when we use Qemu in non toolstack domain for the sake
of Qdisk backend, so xl devd (which spawns Qemu) fails with segmentation
fault afterwards.

See libxl__spawn_qdisk_backend() which explicitly sets:
dmss->guest_config = NULL;

Signed-off-by: Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>
---
 tools/libs/light/libxl_dm.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/tools/libs/light/libxl_dm.c b/tools/libs/light/libxl_dm.c
index e88440f0da..f19f46c296 100644
--- a/tools/libs/light/libxl_dm.c
+++ b/tools/libs/light/libxl_dm.c
@@ -3147,7 +3147,13 @@ static void device_model_spawn_outcome(libxl__egc *egc,
 
     /* Check if spawn failed */
     if (rc) goto out;
-    /* d_config is NULL for xl devd/libxl__spawn_qemu_xenpv_backend(). */
+
+    /*
+     * XXX Maybe it should be: if (!d_config || ...), but
+     * libxl: error: libxl_qmp.c:1334:qmp_ev_lock_aquired:
+     * Domain 2:Failed to connect to QMP socket /var/run/xen/qmp-libxl-2: No such file or directory
+     * Maybe in that case we need to update cmd line args.
+     */
     if (d_config && d_config->b_info.device_model_version
             == LIBXL_DEVICE_MODEL_VERSION_QEMU_XEN) {
         rc = libxl__ev_time_register_rel(ao, &dmss->timeout,
-- 
2.34.1

